channeldb/graph_cache.go

```diff
@@ -0,0 +1,330 @@
+package channeldb
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcutil"
+
+	"github.com/lightningnetwork/lnd/kvdb"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing/route"
+)
+
+// GraphNode is an interface for all the information the cache needs to know
+// about a lightning node.
+type GraphNode interface {
+	// PubKey is the node's public identity key.
+	PubKey() route.Vertex
+
+	// Features returns the node's p2p features.
+	Features() *lnwire.FeatureVector
+
+	// ForEachChannel iterates through all channels of a given node,
+	// executing the passed callback with an edge info structure and the
+	// policies of each end of the channel. The first edge policy is the
+	// outgoing edge *to* the connecting node, while the second is the
+	// incoming edge *from* the connecting node. If the callback returns an
+	// error, then the iteration is halted with the error propagated back up
+	// to the caller.
+	ForEachChannel(kvdb.RTx,
+		func(kvdb.RTx, *ChannelEdgeInfo, *ChannelEdgePolicy,
+			*ChannelEdgePolicy) error) error
+}
+
+// DirectedChannel is a type that stores the channel information as seen from
+// one side of the channel.
+type DirectedChannel struct {
+	// ChannelID is the unique identifier of this channel.
+	ChannelID uint64
+
+	// IsNode1 indicates if this is the node with the smaller public key.
+	IsNode1 bool
+
+	// OtherNode is the public key of the node on the other end of this
+	// channel.
+	OtherNode route.Vertex
+
+	// Capacity is the announced capacity of this channel in satoshis.
+	Capacity btcutil.Amount
+
+	// OutPolicy is the outgoing policy from this node *to* the other node.
+	OutPolicy *ChannelEdgePolicy
+
+	// InPolicy is the incoming policy *from* the other node to this node.
+	InPolicy *ChannelEdgePolicy
```

Comment:
`ChannelEdgePolicy` contains quite a few fields that aren't needed for pathfinding. But I guess that creating a stripped down structure here will increase the PR diff a lot.

Reply:
It does increase the diff quite a bit but also makes it more easy to understand IMO. So I went with a stripped down version which also saved quite a bit of space (I think we're down to 80MB of total memory for a mainnet lnd with one channel).